File: J9Ch/src/J_4_Breadth_First_Search/Related_7/_431_Connected_Component_in_Undirected_Graph_Medium.py
# /**431 Connected Component in Undirected Graph


# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []
class Solution:
    # @param {UndirectedGraphNode[]} nodes a array of undirected graph node
    # @return {int[][]} a connected set of a undirected graph
    def dfs(self, x, tmp):
        self.v[x.label] = True
        tmp.append(x.label)
        for node in x.neighbors:
            if not self.v[node.label]:
                self.dfs(node, tmp)

    def connectedSet(self, nodes):
        # Write your code here
        self.v = {}
        for node in nodes:
            self.v[node.label] = False

        ret = []
        for node in nodes:
            if not self.v[node.label]:
                tmp = []
                self.dfs(node, tmp)
                ret.append(sorted(tmp))
        return ret

File: J9Ch/src/J_4_Breadth_First_Search/Related_7/_431_Connected_Component_in_Undirected_Graph_Medium.cpp
/**431 Connected Component in Undirected Graph */


/**
 * Definition for Undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
map<int, bool> v;
class Solution {
public:
    /**
     * @param nodes a array of Undirected graph node
     * @return a connected set of a Undirected graph
     */
    void dfs(UndirectedGraphNode* x, vector<int> &tmp) {
        v[x->label] = true;
        tmp.push_back(x->label);
        int n = x->neighbors.size();
        for (int i = 0; i < n; ++i)
        if (v.find(x->neighbors[i]->label) == v.end()) {
            dfs(x->neighbors[i], tmp);
        }
    }

    vector<vector<int>> connectedSet(vector<UndirectedGraphNode*>& nodes) {
        // Write your code here
        vector<vector<int>> result;
        int n = nodes.size();
        for (int i = 0; i < n; ++i)
        if (v.find(nodes[i]->label) == v.end()) {
            vector<int> tmp;
            dfs(nodes[i], tmp);
            sort(tmp.begin(), tmp.end());
            result.push_back(tmp);
        }
        return result;
    }
};

